esphome/core/__init__.py

```diff
@@ -712,13 +717,21 @@ def register_variable(self, id, obj):
     def has_id(self, id):
         return id in self.variables
 
+    def reduce_stack_usage(self, stmt):
+        if self.stack_reduction_not_re.search(stmt) is not None:
+            return stmt
+        if self.stack_reduction_re.search(stmt) is not None:
+            return "[&]() __attribute__((noinline)) {" + stmt + "}();"
```

Comment:
```suggestion
            return "[&]() __attribute__((noinline)) {" + stmt + "}();"
```

Hmm, if this replaced a lambda like `bar([=]() { my_local_var.foo(); })` where `my_local_var` is a local variable in the `setup` function I think this might not work.

Because then it will be replaced by

```cpp
[&]() __attribute__((noinline)) { 
  bar([=]() { my_local_var.foo(); });
}();
```

but the outer lambda uses the reference, so the inner one is (I think) illegal.

Anyway, the `[=]` lambda capture was only necessary in very early versions anyway (where all variables were locally defined). I think we can just replace the scope on the normal lambda expression.